File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/rendering/components/A2UIRow.kt
package com.interfaceai.a2ui.rendering.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.interfaceai.a2ui.models.A2UIComponent
import com.interfaceai.a2ui.models.Surface
import com.interfaceai.a2ui.processing.DataBindingResolver
import com.interfaceai.a2ui.processing.TokenResolver
import com.interfaceai.a2ui.rendering.ComponentNodeView
import com.interfaceai.a2ui.rendering.applyA2UIStyle
import kotlinx.serialization.json.JsonElement

@Composable
fun A2UIRowView(
    component: A2UIComponent,
    surface: Surface,
    tokenResolver: TokenResolver,
    dataResolver: DataBindingResolver,
    onAction: ((String, Map<String, JsonElement>?) -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val gap = tokenResolver.resolveStyleFloat(component.style, "gap")
    val childIds = surface.childIds(component)

    Row(
        modifier = modifier.applyA2UIStyle(component.style, tokenResolver),
        horizontalArrangement = Arrangement.spacedBy((gap ?: 0f).dp)
    ) {
        for (childId in childIds) {
            ComponentNodeView(
                componentId = childId,
                surface = surface,
                tokenResolver = tokenResolver,
                dataResolver = dataResolver,
                onAction = onAction
            )
        }
    }
}

File: ios-renderer/Sources/A2UIRenderer/Rendering/Components/A2UIRow.swift
import SwiftUI

/// Renders an A2UI "Row" component as an HStack.
public struct A2UIRowView: View {
    let component: A2UIComponent
    let surface: Surface
    let tokenResolver: TokenResolver
    let dataResolver: DataBindingResolver
    var onAction: ((String, [String: JSONValue]?) -> Void)?

    public var body: some View {
        let gap = tokenResolver.resolveStyleCGFloat(component.style, key: "gap")
        let childIds = surface.childIds(for: component)
        let crossAlign = tokenResolver.resolveStyleString(component.style, key: "crossAxisAlignment")
        let mainAlign = tokenResolver.resolveStyleString(component.style, key: "mainAxisAlignment")

        HStack(alignment: mapVerticalAlignment(crossAlign), spacing: gap ?? 0) {
            if mainAlign == "center" || mainAlign == "max" {
                Spacer(minLength: 0)
            }
            ForEach(childIds, id: \.self) { childId in
                if mainAlign == "space_between" {
                    ComponentNodeView(
                        componentId: childId,
                        surface: surface,
                        tokenResolver: tokenResolver,
                        dataResolver: dataResolver,
                        onAction: onAction
                    )
                    if childId != childIds.last {
                        Spacer(minLength: 0)
                    }
                } else {
                    ComponentNodeView(
                        componentId: childId,
                        surface: surface,
                        tokenResolver: tokenResolver,
                        dataResolver: dataResolver,
                        onAction: onAction
                    )
                }
            }
            if mainAlign == "center" {
                Spacer(minLength: 0)
            }
        }
        .applyA2UIStyle(component.style, tokenResolver: tokenResolver)
    }
}

private func mapVerticalAlignment(_ align: String?) -> VerticalAlignment {
    switch align?.lowercased() {
    case "center": return .center
    case "max": return .bottom
    case "baseline": return .firstTextBaseline
    default: return .top
    }
}
